### PR TITLE
(SERVER-780) Validate node classification

### PR DESCRIPTION
**Pull request by nwolfe**
This commit adds a beaker script that will query the master's HTTP API
and validate that each node will receive the correct classes during the
catalog request.

This beaker script requires that each node already has an SSL
certificate and private key on the master. Unfortunately there is no
automation for putting the certificate and key on the master yet,
see SERVER-787.

This validation script will likely be executed as the very last thing
prior to launching Gatling.

### CODE REVIEW

File: jenkins-integration/config/nodes/example-node1.json [hunk 1]
```diff
@@ -11,14 +11,14 @@
             "path": "/root/test-catalogs/catalog_zero/modules/catalog_zero1"
```
Comment (nwolfe) [inline jenkins-integration/config/nodes/example-node1.json]: These modules are a little easier to work with during development as they have default values for class parameters.
Nothing is using these still, just easier for development.

**Comment by nwolfe** [comment]
Ping @cprice404 @jpinsonault for review.

@cprice404 Querying the catalog endpoint turned out to be the best option. All we need now is the certs/keys for each agent on the master (SERVER-787).

File: jenkins-integration/Gemfile [hunk 2]
```diff
@@ -2,3 +2,7 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
 
 gem 'beaker', '2.15.1'
 gem 'scooter', :git => 'git@github.com:puppetlabs/scooter.git', :tag => '2.2'
+
+if File.exists? "#{__FILE__}.local"
+  eval(File.read("#{__FILE__}.local"), binding)
+end
```
Comment (cprice404) [inline jenkins-integration/Gemfile]: ? this seems spooky

File: jenkins-integration/beaker/install/shared/validate_classification.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    step "Checking classes for node '#{config['certname']}'"
+    catalog_query = build_catalog_query(host, config, server, cacert)
+    response = on(host, catalog_query).stdout.chomp
+    check_for_classes(JSON.parse(response), config)
+  end
+end
+
+scenario_id = ENV['PUPPET_GATLING_SCENARIO']
+nodes = node_configs(scenario_id)
+validate_classification(master, nodes)
```
Comment (cprice404) [inline jenkins-integration/beaker/install/shared/validate_classification.rb]: You put this in `shared`; I guess it works fine for both FOSS and PE?

**Comment by cprice404** [comment]
+1

File: jenkins-integration/Gemfile (hunk 2 above)
Comment (camlow325) [inline jenkins-integration/Gemfile]: FWIW, we do this in Puppet Server, too - https://github.com/puppetlabs/puppet-server/blob/puppet-server-2.1.1/Gemfile#L26-L28.  @jeffmccune probably could give more input here as to when this pattern was established in other Puppet projects.

Comment (nwolfe) [inline jenkins-integration/Gemfile]: Haha this is for loading the `Gemfile.local` if you have one.  Turns out Bundler doesn't have a global file akin to leiningen's `profiles.clj`, so this is how you support developer-specific gems.

File: jenkins-integration/beaker/install/shared/validate_classification.rb (hunk 3 above)
Comment (nwolfe) [inline jenkins-integration/beaker/install/shared/validate_classification.rb]: Yes! Works for both FOSS and PE; just need the client's certificate.

Nothing is calling this script, but I expect the shell script that launches Gatling would probably call this validation first.

**Comment by cprice404** [comment]
@jpinsonault would you mind reviewing/merging this today?

File: jenkins-integration/lib/puppet/gatling/config.rb [hunk 4]
```diff
@@ -21,12 +21,18 @@ def node_configs(scenario_id)
...
+  (env.nil? || env.empty?) ? 'production' : env
+end
+
 # Group the list of node configs into a hash keyed by their environments.
 # A nil or empty environment will be interpreted as 'production'.
 def group_by_environment(node_configs)
   node_configs.group_by do |config|
-    env = config['environment']
-    (env.nil? || env.empty?) ? 'production' : env
+    node_environment(config)
```
Comment (jpinsonault) [inline jenkins-integration/lib/puppet/gatling/config.rb]: Good call, easier to read this way

**Comment by jpinsonault** [comment]
Looks good to me :+1:

**Pull request merged**